src/services/factories/tokenList.ts

```diff
@@ -93,27 +103,119 @@ export function getTokensFactory(factoryParams: {
+        logDebug(`[${networkId}] Address ${tokenAddress} with id ${id} not in local list, fetching`)
+        promises.push(getErc20DetailsOrAddress(networkId, tokenAddress))
+      } else {
+        // Token already exists, update id
+        logDebug(`[${networkId}] Address ${tokenAddress} already in the list, updating id to ${id}`)
+        const token = localAddressesMap.get(tokenAddress) as TokenDetails
+        token.id = id
+      }
+    })
+
+    const partialTokens = await Promise.all(promises)
+
+    // For all newly fetched tokens
+    partialTokens.forEach(partialToken => {
+      if (typeof partialToken === 'string') {
+        // We replaced potential null responses with original tokenAddress string for logging purposes
+        logDebug(`[${networkId}] Address ${partialToken} is not a valid ERC20 token`)
+      } else if (partialToken) {
+        // If we got a valid response
+        logDebug(
+          `[${networkId}] Got details for address ${partialToken.address}: symbol '${partialToken.symbol}' name '${partialToken.name}'`,
+        )
+        // Get id from address/id mapping
+        const id = addressesAndIds.get(partialToken.address) as number
+        // build token object
+        const token: TokenDetails = { ...partialToken, id }
+        // add to local address map
+        localAddressesMap.set(partialToken.address, token)
+      }
+    })
+
+    // Convert map values to a list. Map keeps insertion order, so new tokens are added at the end
+    const tokenList = Array.from(localAddressesMap.values())
+
+    // Persist it \o/
+    tokenListApi.persistTokens({ networkId, tokenList })
+  }
+
+  async function updateTokens(networkId: number): Promise<void> {
+    updatedTokensForNetwork.add(networkId)
 
     try {
-      await updateWithRetries(networkId, tokenListApi.getTokens(networkId))
+      const numTokens = await retry<number>({ fn: exchangeApi.getNumTokens.bind(exchangeApi), fnParams: [networkId] })
+      const tokens = tokenListApi.getTokens(networkId)
+
+      logDebug(`[${networkId}] Contract has ${numTokens}; local list has ${tokens.length}`)
+      if (numTokens > tokens.length) {
+        // When there are more tokens in the contract than locally, fetch the new tokens
+        await updateTokenDetails(networkId, numTokens)
+      } else {
+        // Otherwise, only update the ids
+        await updateTokenIds(networkId, tokens)
+      }
     } catch (e) {
       // Failed to update after retries.
       logDebug(e.message)
       // Clear flag so on next query we try again.
-      updatedIdsForNetwork.delete(networkId)
+      updatedTokensForNetwork.delete(networkId)
     }
   }
 
   return function(networkId: number): TokenDetails[] {
-    if (!updatedIdsForNetwork.has(networkId)) {
-      updateTokenIds(networkId)
+    if (!updatedTokensForNetwork.has(networkId)) {
+      console.log(`[${networkId}] Will update tokens for network`)
```

Comment:
use logDebug? also, add sth like `[tokenListFactory]` ?

Reply:
Good catch. I added them mostly for debugging and thought about removing them later, but I got to like these logs, so I'll make them `logDebug`